sdks/go/pkg/beam/core/runtime/exec/fn.go

```diff
@@ -125,10 +125,144 @@ func Invoke(ctx context.Context, ws []typex.Window, ts typex.EventTime, fn *func
 	return nil, nil
 }
 
+// InvokeWithoutEventTime runs the given function at time 0 in the global window.
 func InvokeWithoutEventTime(ctx context.Context, fn *funcx.Fn, opt *MainInput, extra ...interface{}) (*FullValue, error) {
 	return Invoke(ctx, window.SingleGlobalWindow, mtime.ZeroTimestamp, fn, opt, extra...)
 }
 
+// invoker is a container struct for hot path invocations of DoFns, to avoid
+// repeating fixed set up per element.
+type invoker struct {
+	fn   *funcx.Fn
+	args []interface{}
+	// TODO(lostluck):  2018/07/06 consider replacing with a slice of functions to run over the args slice, as an improvement.
+	ctxIdx, wndIdx, etIdx int   // specialized input indexes
+	outEtIdx, errIdx      int   // specialized output indexes
+	in, out               []int // general indexes
+}
+
+func newInvoker(fn *funcx.Fn) *invoker {
+	n := &invoker{
+		fn:   fn,
+		args: make([]interface{}, len(fn.Param)),
+		in:   fn.Params(funcx.FnValue | funcx.FnIter | funcx.FnReIter | funcx.FnEmit),
+		out:  fn.Returns(funcx.RetValue),
+	}
+	var ok bool
+	if n.ctxIdx, ok = fn.Context(); !ok {
+		n.ctxIdx = -1
+	}
+	if n.wndIdx, ok = fn.Window(); !ok {
+		n.wndIdx = -1
+	}
+	if n.etIdx, ok = fn.EventTime(); !ok {
+		n.etIdx = -1
+	}
+	if n.outEtIdx, ok = fn.OutEventTime(); !ok {
+		n.outEtIdx = -1
+	}
+	if n.errIdx, ok = fn.Error(); !ok {
+		n.errIdx = -1
+	}
+	return n
+}
+
+// ClearArgs zeroes argument entries in the cached slice to allow values to be garbage collected after the bundle ends.
+func (n *invoker) ClearArgs() {
+	for i := range n.args {
+		n.args[i] = nil
+	}
+}
+
+// Invoke invokes the fn with the given values. The extra values must match the non-main
+// side input and emitters. It returns the direct output, if any.
+func (n *invoker) Invoke(ctx context.Context, ws []typex.Window, ts typex.EventTime, opt *MainInput, extra ...interface{}) (*FullValue, error) {
```

Comment:
Consider adding unit tests for the new invoker, maybe something that ensures it's functionally identical to the old one. As we make further changes, it seems easy to forget to update one of them otherwise.

Reply:
Done.

Reply:
I don't see them. Forgot to push a commit?

Reply:
I should have been more explicit. There's presently no difference between the two paths since the "old" invoker doesn't exist anymore. It creates a single use cached invoker, uses it and then returns. The existing tests now cover the cached invoker.

That said, I've added a benchmarked version of the test to compare the "cached" and "single use" approaches, which will be a good direct measure of additional caching work. This also demonstrates whether the cache is doing anything beneficial.